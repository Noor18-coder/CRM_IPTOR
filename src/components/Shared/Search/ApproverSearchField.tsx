/* eslint-disable no-shadow */
import React from 'react';
import { useSelector } from 'react-redux';
import { Typeahead } from 'react-bootstrap-typeahead';
import { AppState } from '../../../store/store';
import { UserItem } from '../../../helpers/Api/models';
import ImageConfig from '../../../config/ImageConfig';

const ApproverRoles = ['Admin', 'Manager'];

interface Props {
  onChange: (data: string) => void;
  description: string;
  currentSelectedUser?: string;
  disabled: boolean;
  value?: string;
}

const ApproverSearchField: React.FC<Props> = (props) => {
  const { onChange, description, disabled, value } = props;
  const state: AppState = useSelector((appState: AppState) => appState);
  const [edit, setEdit] = React.useState<boolean>(false);
  const [options, setOptions] = React.useState<UserItem[]>([]);
  const onUserChange = (user: UserItem[]) => {
    if (user && user.length) {
      setEdit(false);
      onChange(user[0].user);
    }
  };

  const onInputValueChange = (flag: boolean) => {
    setEdit(flag);

    if (flag === true) {
      onChange('');
    }
  };

  React.useEffect(() => {
    const users: UserItem[] = state.users.users.filter((obj: UserItem) => {
      const role = obj.ROLE ? obj.ROLE : '';
      return ApproverRoles.indexOf(role) > -1;
    });
    setOptions(users);
  }, []);

  return (
    <>
      {edit ? (
        <>
          <div>
            <Typeahead
              id="userId"
              onChange={onUserChange}
              multiple={false}
              maxResults={options.length}
              options={options}
              paginate={false}
              placeholder={description}
              autoFocus
              labelKey={(option) => `${option.description}`}
            />
            <button
              type="button"
              className="edit-approver-name"
              onClick={() => {
                onInputValueChange(false);
              }}>
              <img src={ImageConfig.CLOSE_BTN} alt="edit" />
            </button>
          </div>
        </>
      ) : (
        <>
          <input type="text" value={`${value}`} disabled={disabled} className="form-control" />
          <button
            type="button"
            className="edit-approver-name"
            onClick={() => {
              onInputValueChange(true);
            }}>
            <img src={ImageConfig.EDIT_ICON} alt="edit" />
          </button>
        </>
      )}
    </>
  );
};

export default ApproverSearchField;
