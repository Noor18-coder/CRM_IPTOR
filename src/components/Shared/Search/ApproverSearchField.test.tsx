import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import ApproverSearchField from './ApproverSearchField';
import StoreMock from '../../../mocks/Store.mock';

const onChange = jest.fn();
const description = 'some description';
const disabled = false;
const currentSelectedUser = 'user';
const value = 'theValue';

const store = StoreMock.createInitialStore();

describe('[Shared] ApproverSearchField', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ApproverSearchField
            onChange={onChange}
            description={description}
            disabled={disabled}
            currentSelectedUser={currentSelectedUser}
            value={value}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
