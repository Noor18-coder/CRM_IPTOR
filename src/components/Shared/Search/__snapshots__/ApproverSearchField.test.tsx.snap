// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Shared] ApproverSearchField should renders correctly 1`] = `
Array [
  <input
    className="form-control"
    disabled={false}
    type="text"
    value="theValue"
  />,
  <button
    className="edit-approver-name"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="edit"
      src="edit-icon.svg"
    />
  </button>,
]
`;
